skiplang/prelude/src/stdlib/collections/persistent/Queue.sk

```diff
@@ -0,0 +1,126 @@
+module Queue;
+
+/* A persistent First-In First-Out queue */
+class .Queue<+T> private {
+  private siz: Int,
+  private head: List<T>,
+  private reversedTail: List<T>,
+} extends Sequence<T> uses Equality[T: Equality], Orderable[T: Orderable] {
+  /* Invariant:
+     - this.size() == this.head.size() + this.reversedTail.size()
+     - The whole queue is the concatenation of head and reversedTail.reversed()
+   */
+  static fun empty(): this {
+    static{siz => 0, head => List.Nil(), reversedTail => List.Nil()}
+  }
+
+  static fun createFromItems<C: readonly Sequence<T>>(items: C): Queue<T> {
+    static{
+      siz => items.size(),
+      head => List.Nil(),
+      reversedTail => List::reverseFromIterator(items.values()),
+    }
+  }
+
+  fun size(): Int {
+    this.siz
+  }
+
+  fun push<U>[T: U](x: U): Queue<U> {
+    Queue<U>{
+      siz => this.siz + 1,
+      head => this.head,
+      reversedTail => List.Cons(x, this.reversedTail),
+    }
+  }
+
+  fun pop(): ?(this, T) {
+    this match {
+    | Queue{head => List.Cons(res, head), siz} ->
+      Some((this with {head, siz => siz - 1}, res))
+    | Queue{head => List.Nil(), reversedTail => List.Nil(), siz} ->
+      invariant(siz == 0);
+      None()
+    | Queue{head => List.Nil(), reversedTail => List.Cons(rhd, rtl), siz} ->
+      rtl.revAppend(List::of(rhd)) match {
```

Comment:
Is this destructuring/restructuring of reversedTail necessary?  i.e. don't we know (from the fact that we passed through the previous case) that reversedTail isn't Nil?

```suggestion
    | Queue{head => List.Nil(), reversedTail, siz} ->
      reversedTail.reversed() match {
```

Reply:
You're right, though `reversedTail.reversed()` returns a `List`, not a `Cons`, and I didn't manage to change `reversed` signature so that `Cons(_).reversed()` is a `Cons`, and I hate `_ -> invariant_violation` because the best way to be sure it doesn't happen is when it's not written